src/Features/LanguageServer/Protocol/Extensions/Extensions.cs

```diff
@@ -59,16 +74,14 @@ public static Uri GetUriFromProjectPath(this TextDocument document)
             return ProtocolConversions.GetUriFromFilePath(path);
         }
 
-        public static Uri? TryGetURI(this TextDocument document, RequestContext? context = null)
-            => ProtocolConversions.TryGetUriFromFilePath(document.FilePath, context);
-
-        public static ImmutableArray<Document> GetDocuments(this Solution solution, Uri documentUri)
+        public static async ValueTask<ImmutableArray<Document>> GetDocumentsAsync(this Solution solution, Uri documentUri, CancellationToken cancellationToken)
         {
             var documentIds = GetDocumentIds(solution, documentUri);
 
-            // We don't call GetRequiredDocument here as the id could be referring to an additional document.
-            var documents = documentIds.Select(solution.GetDocument).WhereNotNull().ToImmutableArray();
-            return documents;
+            // We don't call GetTextDocumentAsync here as the id could be referring to an additional document.
+            var documents = await documentIds.SelectAsArrayAsync(
+                static (id, solution, ct) => solution.GetDocumentAsync(id, includeSourceGenerated: true, ct), solution, cancellationToken).ConfigureAwait(false);
+            return documents.WhereNotNull().ToImmutableArray();
```

Comment:
seems odd that we always force source generation here.  woudln't we know based on the uri if we should get SG docs or not?

Reply:
It shouldn't actually force the source generation: the GetDocumentAsync() method fast-paths if the DocumentId is a regular document in the solution, and GetDocumentIds will have found and returned that first.

Reply:
I think it should be fine, but in the case where the document doesn't exist and it goes to misc it could cause source generation.  But I think that is OK since it won't happen often?

Reply:
> it goes to misc it could cause source generation. But I think that is OK since it won't happen often?

Yes, it'd run in that case, but we'd have to run to know there isn't a document either way if we want that to be fully refreshed. This is quite niche though so I'm not too worried -- we've got bigger fish to fry on the source generator performance space.